zap/src/main/java/org/zaproxy/zap/spider/SpiderController.java

```diff
@@ -299,107 +321,51 @@ public void resourceURIFound(
-                        HttpRequestHeader.GET);
-        spider.submitTask(task);
-    }
-
-    @Override
-    public void resourceURIFound(HttpMessage responseMessage, int depth, String uri) {
-        resourceURIFound(responseMessage, depth, uri, false);
-    }
-
-    @Override
-    public void resourcePostURIFound(
-            HttpMessage responseMessage, int depth, String uri, String requestBody) {
-        log.debug("New POST resource found: " + uri);
-
-        // Check if the uri was processed already
-        synchronized (visitedPost) {
-            if (arrayKeyValueExists(uri, requestBody)) {
-                log.debug("URI already visited: " + uri);
-                return;
-            } else {
-                if (visitedPost.containsKey(uri)) {
-                    visitedPost.get(uri).add(requestBody);
-                } else {
-                    ArrayList<String> l = new ArrayList<String>();
-                    l.add(requestBody);
-                    visitedPost.put(uri, l);
-                }
-            }
-        }
-
-        // Create the uri
-        URI uriV = createURI(uri);
-        if (uriV == null) {
-            return;
-        }
-
-        // Check if any of the filters disallows this uri
-        for (FetchFilter f : fetchFilters) {
-            FetchStatus s = f.checkFilter(uriV);
-            if (s != FetchStatus.VALID) {
-                log.debug("URI: " + uriV + " was filtered by a filter with reason: " + s);
-                spider.notifyListenersFoundURI(uri, HttpRequestHeader.POST, s);
-                return;
-            }
-        }
-
-        spider.notifyListenersFoundURI(uri, HttpRequestHeader.POST, FetchStatus.VALID);
-
-        // Submit the task
-        SpiderTask task =
-                new SpiderTask(
-                        spider,
-                        responseMessage.getRequestHeader().getURI(),
-                        uriV,
-                        depth,
-                        HttpRequestHeader.POST,
-                        requestBody);
+        SpiderTask task = new SpiderTask(spider, resourceFound, uriV);
         spider.submitTask(task);
     }
 
     /**
      * Checks whether the value exists in an ArrayList of certain key.
      *
-     * @param key the string of the uri
-     * @param value the request body of the uri
-     * @return true or false depending whether the uri and request body have already been processed
+     * @param headers list of HTTP headers
+     * @return canonical string representation of headers
      */
-    private boolean arrayKeyValueExists(String key, String value) {
-        if (visitedPost.containsKey(key)) {
-            for (String s : visitedPost.get(key)) {
-                if (s.equals(value)) {
-                    return true;
-                }
-            }
+    private String getCanonicalHeadersString(List<HttpHeaderField> headers) {
+        if (headers != null && !headers.isEmpty()) {
+            headers.removeIf(
```

Comment:
Better do these validations when the resource is created, here it should just be used.

Reply:
Resolved via last commit, validations are now done as part of the builder